2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,517 @@
+  var slash = Slash{epoch: evidence.epoch, rate: 0, vpower_fraction: compute_vpower_fraction(evidence.type, vpower)}
+  //enqueue slash (Step 1.1 of cubic slashing)
+  append(enqueued_slashes[evidence.epoch + unbonding_length], slash)
+  //jail validator (Step 1.2 of cubic slashing)
+  validators[validator_address].state[cur_epoch + 1] = jailed
+  remove_from_validator_sets(validator_address, cur_epoch + 1)
+  //freeze validator to prevent delegrators from altering their delegations (Step 1.3 of cubic slashing)
+  validators[validator_address].frozen = true
+}
+```
+
+```go
+func update_voting_power(validator_address, epoch)
+{
+  //compute bonds from total_deltas 
+  var bonds = read_epoched_field(validators[validator_address].total_deltas, epoch, 0)
+  //compute the new voting power
+  var power_after = votes_per_token*bonds
+  //update voting power and return it
+  validators[validator_address].voting_power[epoch] = power_after
+  return power_after
+}
+```
+```go
+func update_total_voting_power(epoch)
+{
+  var total = 0
+  forall (validator in validator_sets[epoch].active U validator_sets[epoch].inactive) do
+    total += validator.voting_power
+  total_voting_power[epoch] = total
+}
+```
+```go
+func update_validator_sets(validator_address, epoch, power_before, power_after)
+{
+  var min_active = first(validator_sets[epoch].active)
+  var max_inactive = last(validator_sets[epoch].inactive)
+  if (power_before >= max_inactive.voting_power) then
+    //if active but it loses power below the max_inactive then move validator to inactive
+    //and promote max_inactive
+    if (power_after < max_inactive.voting_power) then
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, max_inactive)
+      remove(validator_sets[epoch].inactive, max_inactive.validator)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+    //if active and its change in power does not degrade it to inactive, then update its voting power
+    else
+      remove(validator_sets[epoch].active, validator_address)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+  else
+    //if inactive and gains power above the min_active, then insert into active and
+    //degrade min_active 
+    if (power_after > min_active.voting_power) then
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, min_active)
+      remove(validator_sets[epoch].active, min_active.validator)
+      add(validator_sets[epoch].active, <validator_address, power_after>)
+    //if inactive and its change in power does not promote it to active, then update its voting power
+    else
+      remove(validator_sets[epoch].inactive, validator_address)
+      add(validator_sets[epoch].inactive, <validator_address, power_after>)
+}
+```
+
+```go
+func remove_from_validator_sets(validator_address)
+{
+  var max_inactive = last(validator_sets[epoch].inactive)
```

Comment:
where is `epoch` defined? Is it missing as parameter?

Reply:
yes, It was missing a parameter.